src/storage/txn/scheduler.rs

```diff
@@ -11,92 +11,508 @@
+                                Ok(result.drain(..).map(|x| x.map_err(::storage::Error::from)).collect())
+                            }
+                            _ => { panic!("prewrite return but process result is not result set."); }
+                        }
+                    }
+                    Err(e) => Err(e.into())
+                });
+            }
+            Command::CommitThenGet { ref mut callback, .. } => {
+                callback.take().unwrap()(match result {
+                    Ok(()) => {
+                        match pr {
+                            ProcessResult::Value { value } => { Ok(value) }
+                            _ => { panic!("commit then get return but process result is not value."); }
+                        }
+                    }
+                    Err(e) => {
+                        Err(::storage::Error::from(e))
+                    }
+                });
+            }
+            Command::Commit { ref mut callback, .. } |
+            Command::Cleanup { ref mut callback, .. } |
+            Command::Rollback { ref mut callback, .. } => {
+                callback.take().unwrap()(result.map_err(::storage::Error::from));
+            }
+            Command::RollbackThenGet { ref mut callback, .. } => {
+                callback.take().unwrap()(match result {
+                    Ok(()) => {
+                        match pr {
+                            ProcessResult::Value { value } => { Ok(value) }
+                            _ => { panic!("rollback then get return but process result is not value."); }
+                        }
+                    }
+                    Err(e) => {
+                        Err(::storage::Error::from(e))
+                    }
+                });
+            }
+            _ => { panic!("unsupported write cmd"); }
         }
-        Command::Commit { ctx, keys, lock_ts, commit_ts, callback } => {
-            callback(store.commit(ctx, keys, lock_ts, commit_ts)
-                .map_err(::storage::Error::from));
+    }
+
+    fn finish_cmd(&mut self, cid: u64) {
+        let ctx = self.cmd_ctxs.remove(&cid).unwrap();
+        assert_eq!(cid, ctx.cid);
+
+        if ctx.needed_latches().is_empty() {
+            return;
         }
-        Command::CommitThenGet { ctx, key, lock_ts, commit_ts, get_ts, callback } => {
-            callback(store.commit_then_get(ctx, key, lock_ts, commit_ts, get_ts)
-                .map_err(::storage::Error::from));
+
+        let wakeup_list = self.latches.release(ctx.needed_latches(), cid);
+        for wcid in wakeup_list {
+            self.wakeup_cmd(wcid);
         }
-        Command::Cleanup { ctx, key, start_ts, callback } => {
-            callback(store.cleanup(ctx, key, start_ts).map_err(::storage::Error::from));
+    }
+
+    fn wakeup_cmd(&mut self, cid: u64) {
+        let all_acquired = self.acquire_latches(cid);
+        if all_acquired {
+            self.get_snapshot(cid);
```

Comment:
All commands need a snapshot for read. But I think we may reuse some old snapshots if we really need to do it later.